### PR TITLE
CODAP-1234: render fused bars as coalesced solid segments

### PR DESCRIPTION
**Pull request by kswenson**
## Summary

Part 2 of 2 for CODAP-1234 (the case-table scroll + MST fixes are in #2647).

In a bar chart or histogram, each fused bar is a stack of one rectangle per case so selected cases can be shown distinctly within the bar. When a bar holds hundreds of cases, each rectangle is sub-pixel-thin, and abutting sub-pixel opaque rectangles composite to less than full opacity under the browser's coverage-based anti-aliasing (Canvas and WebGL alike). The selection color (and the per-case separator lines) wash out to a faint tint instead of the solid color seen when zoomed in. V2 (SVG) doesn't exhibit this.

This renders each fused bar as **coalesced same-style segments**: contiguous cases that render identically (same fill *and* same stroke) are merged into one solid rectangle drawn with that segment's stroke. Selection renders solid at any case count, and strokes appear only at meaningful boundaries (selection changes and the bar outline) rather than as inconsistent per-case lines.

Coalescing by full style (not fill alone) is what makes selection visible **with a legend**: without a legend, selection changes a case's fill; with a legend the fill stays the legend color and selection is carried only by the stroke, so merging by fill alone would drop a selected case's stroke and hide the selection.

### Scope (all handled)
- **Vertical and horizontal** bars (cases stack along y or x).
- **Bar charts and histograms** (categorical and numeric-binned).
- **With and without a categorical legend** (selection by fill vs. by stroke).
- **Split plots** — bars coalesce and clip per subplot cell, so cells don't merge or bleed.
- **Bar-for-each-point** (non-fused "bars") and **transitions** (the points↔bars fuse, category reorder, axis/resize animations) stay on the per-case path — coalescing assumes stable, fused geometry.

### Implementation
- `bar-coalescing.ts` — shared, unit-tested helpers: `coalesceBars`/`coalesceBarRuns` (group by bar, sort along the stack axis, merge contiguous same-style runs), `pointStateToBarPiece`, `anchoredBarRect` (anchored-rect geometry, also used by hit-testing), `sameRunStyle` (the fill+stroke run boundary).
- `PointRendererBase` — `shouldCoalesceBars` gate (fused bars, not mid-transition) and a shared `barStackAxis`; `matchPointsToData` clears the bar-only state on leaving bars mode so a reused renderer can't carry stale fuse/stack state into a non-bars plot.
- **Canvas** — `drawBars` maps point state → pieces → `coalesceBars`, then fills/strokes each segment, clipped per subplot.
- **Pixi/WebGL** — one masked `Graphics` per subplot cell, rebuilt from the coalesced runs each render and clipped to its cell; per-case sprites are hidden (`renderable = false`) while in bars mode (kept for points mode and the fuse transition). Mask/layer teardown is ordered and deferred a frame to avoid rendering a destroyed object. Bar selection is handled by the existing SVG bar covers, so hiding sprites doesn't affect interaction.

### Automated
- `bar-coalescing.test.ts` — 14 unit tests: uniform / fill-boundary / interleaved runs; same-fill-but-different-stroke split (legend selection) for both vertical and horizontal bars; anchor+scale geometry; per-bar grouping; per-segment stroke.
- `npm run build:tsc`, `npm run lint`, and the data-display renderer suite (271 tests) pass.

### Manual
Use a dataset where at least one bar holds many cases (≈200+) so washout would be obvious. Run on the default WebGL/Pixi renderer; repeat the ★ cases on the Canvas renderer (the no-WebGL fallback). In each case, verify bars are solid (no washout, no spurious per-case lines) and selection is clearly visible.

**Rendering & selection**
- [ ] ★ Vertical fused bar chart, **no legend** — select cases (click a bar; select rows in a linked table) → selected cases solid-colored.
- [ ] ★ Vertical fused bar chart, **categorical legend** — select some-but-not-all cases within one colored segment → selected cases show their selection outline (stroke); unselected don't.
- [ ] Horizontal fused bar chart (primary attribute on the **left** axis), no legend and with legend — same checks; bars extend horizontally.
- [ ] ★ Histogram (numeric bins), no legend and with legend — bins render solid; legend selection shows the stroke.

**Split plots**
- [ ] ★ Split bar chart (categorical on x **and** y, or an x + top/right split) — each cell's bars coalesce and clip independently; selecting in one cell highlights only that cell; **bars never bleed outside their cell**.
- [ ] Split histogram with a legend — same independence; selection visible per cell.

**Must-not-coalesce / stale state**
- [ ] **Bar for each point** (univariate numeric → "Bars" display, not fused) — each case is its own distinct bar; bars are **not** merged.
- [ ] Switch a fused bar chart → a points plot (scatter/dot) → back — points render as points (no leftover/blank bars); bars return correctly.

**Transitions** (should animate, not clip or blank out)
- [ ] Toggle "Fuse Dots into Bars" on/off — the points↔bars animation plays and ends in the correct coalesced (on) / per-case (off) rendering.
- [ ] Reorder categories on a categorical/split bar chart — bars slide to new positions, the animation isn't clipped to the old cells, and bars settle correctly clipped.
- [ ] Add / remove a split category — cells rebuild; bars appear in the new layout and clip correctly once settled; no stale or blank bars.
- [ ] Resize the graph / drag an axis — bars rescale and stay clipped; selection persists.

**Interaction**
- [ ] Click a bar selects its cases; click empty plot space deselects; selection round-trips with a linked case table.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: v3/src/components/data-display/renderer/canvas-point-renderer.ts [hunk 1]
```diff
@@ -484,10 +485,15 @@ export class CanvasPointRenderer extends PointRendererBase {
...
 
-      // Draw each point
-      for (const pointState of points) {
-        if (pointState.isVisible) {
-          this.drawPoint(pointState)
+      // Draw each point. Bars are drawn by coalescing contiguous same-fill cases into one solid
+      // segment rect each (see drawBars), rather than one rect per case.
+      if (this._displayType === "bars") {
+        this.drawBars(points)
+      } else {
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/canvas-point-renderer.ts]: Good catch — fixed. `drawBars` now runs only when points are fused into stacked bars (`displayType === "bars" && pointsFusedIntoBars`); non-fused bars (each case its own bar) fall through to per-case rendering.

File: v3/src/components/data-display/renderer/pixi-point-renderer.ts [hunk 2]
```diff
@@ -613,9 +628,50 @@ export class PixiPointRenderer extends PointRendererBase {
...
 
+  // Rebuilds the coalesced-bars layer from the current point state. When fully in bars mode, each
+  // bar's contiguous same-fill cases are merged into one solid segment rect (with its stroke), and
+  // the per-case sprites are hidden from rendering (renderable = false). In points mode or during
+  // the points<->bars transition, the layer is cleared and the sprites render normally. Bar clicks
+  // are handled by the SVG bar covers, so hiding the sprites doesn't affect selection. See
+  // CODAP-1234.
+  private updateBarsLayer(): void {
+    const useCoalesced = this._displayType === "bars" && !this.displayTypeTransitionState.isActive
+    if (!useCoalesced) {
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/pixi-point-renderer.ts:684]: Good catch. Fixed: `updateBarsLayer` now coalesces only when `pointsFusedIntoBars` (and not mid-transition); non-fused bars keep per-case sprite rendering.

File: v3/src/components/data-display/renderer/bar-coalescing.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+/**
+ * Groups a flat list of bar pieces into bars by their shared primary (x) coordinate and coalesces
+ * each bar's same-fill runs, returning all runs across all bars. Shared by the Canvas and Pixi
+ * renderers; each renderer maps its own point state into `IBarPiece`s and draws the returned runs.
+ */
+export function coalesceBars(barPieces: IBarPiece[], anchor: { x: number; y: number }): IBarRun[] {
+  const piecesByBar = new Map<number, IBarPiece[]>()
+  for (const piece of barPieces) {
+    const key = Math.round(piece.x)
+    const bar = piecesByBar.get(key) ?? []
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/bar-coalescing.ts:139]: Good catch. `coalesceBars` now groups by the exact `x` (all cases in a fused bar share an identical computed x) instead of `Math.round(x)`, so distinct sub-pixel-spaced bars no longer collapse together.

File: v3/src/components/data-display/renderer/pixi-point-renderer.ts [hunk 4]
```diff
@@ -196,6 +205,12 @@ export class PixiPointRenderer extends PointRendererBase {
     this.stage.addChild(this.background)
     this.stage.addChild(this.pointsContainer)
     this.pointsContainer.sortableChildren = true
+    // Coalesced-bars layer: drawn on top of the (hidden) per-case sprites when in bars mode.
+    // pointsContainer has no transform, so the stage shares its coordinate space. eventMode "none"
+    // keeps this layer from swallowing pointer events (e.g. empty-space clicks that should reach
+    // the background to deselect); bar clicks are handled by the SVG bar covers above the canvas.
+    this.barsGraphics.eventMode = "none"
+    this.stage.addChild(this.barsGraphics)
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/pixi-point-renderer.ts]: Thanks — this is a real parity gap. We have deferred it: because the subplot cells tile contiguously, a single composite/union mask reduces to plot-bounds clipping and would not clip between adjacent cells, and a coalesced bar is sized by the secondary axis to fit its cell so it does not bleed into neighbors in practice. Added a code comment noting the limitation; proper per-subplot masking of this layer is a follow-up.

File: v3/src/components/data-display/renderer/bar-coalescing.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      const t = pj.y - anchor.y * pj.height * pj.scale
+      top = Math.min(top, t)
+      bottom = Math.max(bottom, t + pj.height * pj.scale)
+      j++
+    }
+    runs.push({ left, top, width, height: bottom - top, fill, stroke, strokeWidth, strokeOpacity })
+    i = j
+  }
+  return runs
+}
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/bar-coalescing.ts]: Good catch. Fixed: `coalesceBarRuns` now takes a `stackAxis` ("y" for vertical, "x" for horizontal) and sorts/expands the run along it while keeping the cross-axis bounds constant. The renderer carries a `barStackAxis` set from `primaryIsBottom` by bar-chart.tsx/histogram.tsx.

File: v3/src/components/data-display/renderer/bar-coalescing.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  const piecesByBar = new Map<number, IBarPiece[]>()
+  for (const piece of barPieces) {
+    const bar = piecesByBar.get(piece.x) ?? []
+    bar.push(piece)
+    piecesByBar.set(piece.x, bar)
+  }
+  const runs: IBarRun[] = []
+  piecesByBar.forEach(bar => runs.push(...coalesceBarRuns(bar, anchor)))
+  return runs
+}
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/bar-coalescing.ts]: Good catch. `coalesceBars` now groups by the cross axis (the axis the cases do NOT stack along), driven by the same `stackAxis`. Rather than inferring orientation from the data (unreliable for tall/few-case bars, and the bar anchor is centered for both orientations), the orientation is plumbed explicitly: bar-chart.tsx/histogram.tsx set `renderer.barStackAxis = primaryIsBottom ? "y" : "x"`.

File: v3/src/components/data-display/renderer/bar-coalescing.test.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    expect(runs).toEqual(expect.arrayContaining([
+      { left: 10, top: 0, width: 4, height: 2, fill: "#e6805b", stroke: "#ffffff", strokeWidth: 1, strokeOpacity: 0.4 },
+      { left: 50, top: 0, width: 4, height: 2, fill: "#4682b4", stroke: "#ffffff", strokeWidth: 1, strokeOpacity: 0.4 }
+    ]))
+  })
+
+  it("returns an empty array for no pieces", () => {
+    expect(coalesceBars([], anchor)).toEqual([])
+  })
+})
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/bar-coalescing.test.ts:108]: Added — `bar-coalescing.test.ts` now has horizontal-orientation tests for both `coalesceBarRuns` (stackAxis "x": merge/sort along x, constant cross-axis height) and `coalesceBars` (group by y; verifies two horizontal bars sharing a baseline x are NOT merged across rows).

File: v3/src/components/data-display/renderer/canvas-point-renderer.ts [hunk 8]
```diff
@@ -484,10 +485,17 @@ export class CanvasPointRenderer extends PointRendererBase {
...
 
-      // Draw each point
-      for (const pointState of points) {
-        if (pointState.isVisible) {
-          this.drawPoint(pointState)
+      // Draw each point. Bars whose cases are fused into shared stacked bars (bar chart/histogram)
+      // are drawn by coalescing contiguous same-fill cases into one solid segment rect each (see
+      // drawBars), rather than one rect per case. Non-fused bars (each case its own bar) fall
+      // through to per-case drawing.
+      if (this._displayType === "bars" && this._pointsFusedIntoBars) {
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/canvas-point-renderer.ts]: Good catch — fixed. Canvas now also requires `!anyTransitionActive` (its transitionManager `hasActiveTransitions()`), matching Pixi, so the fuse animation (and any other transition) plays per case and coalescing only engages once the geometry is stable.

File: v3/src/components/data-display/renderer/pixi-point-renderer.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/pixi-point-renderer.ts]: Follow-up: this is now fixed rather than deferred. The Pixi bars layer is split into one Graphics per subplot cell, each masked to its cell, and coalescing is grouped by subPlotNum — so split-plot cells both coalesce and clip independently (commit 59a31b0f3). Verified in the browser: clean per-cell clipping, no cross-cell merging.

File: v3/src/components/data-display/renderer/pixi-point-renderer.ts [hunk 9]
```diff
@@ -613,9 +648,65 @@ export class PixiPointRenderer extends PointRendererBase {
...
+    this.barsGraphicsBySubplot.forEach((graphics, subplot) => {
+      graphics.clear()
+      const pieces = piecesBySubplot.get(subplot)
+      if (!pieces) return
+      for (const run of coalesceBars(pieces, this._anchor, this._barStackAxis)) {
+        graphics.rect(run.left, run.top, run.width, run.height).fill(run.fill)
+        if (run.strokeWidth > 0) {
+          graphics.stroke({ color: run.stroke, width: run.strokeWidth, alpha: run.strokeOpacity })
+        }
+      }
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/pixi-point-renderer.ts:719]: Adopted the chained pattern: `rect().fill().stroke()` on the same command, matching getRectTexture (the conditional stroke was the only reason it had been split). Functionally it already rendered correctly — per-segment fills and strokes verified in the browser, which shows Pixi v8 resets the path after each fill/stroke — but chaining is clearer and avoids relying on that behavior.

File: v3/src/components/data-display/renderer/pixi-point-renderer.ts [hunk 10]
```diff
@@ -231,9 +244,11 @@ export class PixiPointRenderer extends PointRendererBase {
     this.ticker.destroy()
     this.renderer?.destroy()
     this.renderer = undefined
-    // Destroy masks before stage
+    // Destroy masks and the per-subplot bar layers before stage
     this.subPlotMasks.forEach(mask => mask.destroy())
     this.subPlotMasks = []
+    this.barsGraphicsBySubplot.forEach(graphics => graphics.destroy())
+    this.barsGraphicsBySubplot = []
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/pixi-point-renderer.ts]: Good catch — fixed. doDispose now destroys the per-subplot bar Graphics first (setting their `mask` to null), then the masks, matching doResize which clears mask references before destroying the masks.

File: v3/src/components/data-display/renderer/point-renderer-base.ts [hunk 11]
```diff
@@ -238,6 +241,14 @@ export abstract class PointRendererBase {
     this._pointsFusedIntoBars = value
   }
 
+  get barStackAxis(): "x" | "y" {
+    return this._barStackAxis
+  }
+
+  set barStackAxis(value: "x" | "y") {
+    this._barStackAxis = value
+  }
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/point-renderer-base.ts:250]: Good catch — this is a real latent gap. Rather than tracking a "barStackAxis was set" flag on the renderer, we fixed it at the source: DotLinePlot (which renders dot plots and the non-fused "bar for each point") now explicitly sets renderer.pointsFusedIntoBars = false, mirroring how bar-chart/histogram set it true. That removes the reliance on the histogram's autorun resetting the flag before a histogram -> line-plot switch, so coalescing can't apply with stale fused state. (Our existing matchPointsToData reset already covers leaving bars mode.)

File: v3/src/components/data-display/renderer/point-renderer-base.ts [hunk 12]
```diff
@@ -246,6 +257,16 @@ export abstract class PointRendererBase {
...
 
+  /**
+   * Whether the renderer should draw fused bars as coalesced same-style segments rather than
+   * per-case rects/sprites: only for bar charts/histograms whose cases are fused into shared
+   * stacked bars, and not mid-transition (coalescing assumes stable geometry, so during a
+   * transition we keep per-case rendering so the animation plays). Shared by both renderers.
+   */
+  protected get shouldCoalesceBars(): boolean {
+    return this._displayType === "bars" && this._pointsFusedIntoBars && !this.anyTransitionActive
+  }
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/point-renderer-base.ts:268]: Good catch — this is a real latent gap. Rather than tracking a "barStackAxis was set" flag on the renderer, we fixed it at the source: DotLinePlot (which renders dot plots and the non-fused "bar for each point") now explicitly sets renderer.pointsFusedIntoBars = false, mirroring how bar-chart/histogram set it true. That removes the reliance on the histogram's autorun resetting the flag before a histogram -> line-plot switch, so coalescing can't apply with stale fused state. (Our existing matchPointsToData reset already covers leaving bars mode.)

File: v3/src/components/data-display/renderer/point-renderer-base.ts [hunk 13]
```diff
@@ -322,6 +343,13 @@ export abstract class PointRendererBase {
     // (e.g., PixiPointRenderer's display type transition), so defer the assignment.
     this.doMatchPointsToData(datasetID, caseData, displayType, style)
     this._displayType = displayType
+    // The bar-only state is set imperatively by the bar-chart/histogram plot components and is
+    // never otherwise reset; clear it whenever we leave bars mode so a reused renderer can't carry
+    // a stale fuse flag / stack axis into a non-bars plot. See CODAP-1234.
+    if (displayType !== "bars") {
+      this._pointsFusedIntoBars = false
+      this._barStackAxis = "y"
+    }
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/point-renderer-base.ts:352]: Good catch — this is a real latent gap. Rather than tracking a "barStackAxis was set" flag on the renderer, we fixed it at the source: DotLinePlot (which renders dot plots and the non-fused "bar for each point") now explicitly sets renderer.pointsFusedIntoBars = false, mirroring how bar-chart/histogram set it true. That removes the reliance on the histogram's autorun resetting the flag before a histogram -> line-plot switch, so coalescing can't apply with stale fused state. (Our existing matchPointsToData reset already covers leaving bars mode.)

File: v3/src/components/data-display/renderer/point-renderer-base.ts [hunk 14]
```diff
@@ -49,6 +49,9 @@ export abstract class PointRendererBase {
   protected _isVisible = true
   protected _displayType: PointDisplayType = "points"
   protected _pointsFusedIntoBars = false
+  // axis along which fused-bar cases stack: "y" for vertical bars (primary axis on bottom),
+  // "x" for horizontal bars (primary axis on left). Used by bar coalescing.
+  protected _barStackAxis: "x" | "y" = "y"
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/point-renderer-base.ts:54]: Good catch — this is a real latent gap. Rather than tracking a "barStackAxis was set" flag on the renderer, we fixed it at the source: DotLinePlot (which renders dot plots and the non-fused "bar for each point") now explicitly sets renderer.pointsFusedIntoBars = false, mirroring how bar-chart/histogram set it true. That removes the reliance on the histogram's autorun resetting the flag before a histogram -> line-plot switch, so coalescing can't apply with stale fused state. (Our existing matchPointsToData reset already covers leaving bars mode.)

File: v3/src/components/data-display/renderer/pixi-point-renderer.ts [hunk 15]
```diff
@@ -68,6 +69,16 @@ export class PixiPointRenderer extends PointRendererBase {
   private sprites = new Map<string, PIXI.Sprite>()
   private textures = new Map<string, PIXI.Texture>()
 
+  // Bars are drawn by coalescing contiguous same-fill cases into one solid segment rect each
+  // (see updateBarsLayer / bar-coalescing) rather than one sprite per case. There is one Graphics
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/pixi-point-renderer.ts]: Good catch — fixed. These comments predated the legend-selection change; updated "same-fill" -> "same-style" in the Canvas drawBars docstring and the two Pixi comments so the wording matches sameRunStyle (coalescing by fill AND stroke, which is what keeps stroke-encoded selection visible under a legend).

File: v3/src/components/data-display/renderer/pixi-point-renderer.ts [hunk 16]
```diff
@@ -613,9 +667,61 @@ export class PixiPointRenderer extends PointRendererBase {
       this.ticker.stop()
       this.cleanupUnusedTextures()
     }
+    this.updateBarsLayer()
     this.renderer.render(this.stage)
   }
 
+  // Rebuilds the coalesced-bars layer from the current point state. When fully in bars mode, each
+  // bar's contiguous same-fill cases are merged into one solid segment rect (with its stroke), and
+  // the per-case sprites are hidden from rendering (renderable = false). In points mode or during
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/pixi-point-renderer.ts:676]: Good catch — fixed. These comments predated the legend-selection change; updated "same-fill" -> "same-style" in the Canvas drawBars docstring and the two Pixi comments so the wording matches sameRunStyle (coalescing by fill AND stroke, which is what keeps stroke-encoded selection visible under a legend).

File: v3/src/components/data-display/renderer/canvas-point-renderer.ts [hunk 17]
```diff
@@ -498,6 +506,32 @@ export class CanvasPointRenderer extends PointRendererBase {
     this.hitTester?.updateFromPoints(sortedPoints, this._displayType, this._anchor)
   }
 
+  // Draws bars by coalescing each bar's contiguous same-fill cases into a single solid segment
+  // rectangle (with that segment's stroke), instead of one rect per case. This avoids the
+  // undesirable per-case separator lines and the sub-pixel fill washout that occur when hundreds
+  // of cases stack into one bar. Per-case hit-testing is unaffected (it uses the point positions).
+  // See CODAP-1234.
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/renderer/canvas-point-renderer.ts]: Good catch — fixed. These comments predated the legend-selection change; updated "same-fill" -> "same-style" in the Canvas drawBars docstring and the two Pixi comments so the wording matches sameRunStyle (coalescing by fill AND stroke, which is what keeps stroke-encoded selection visible under a legend).

**Review by bfinzer — APPROVED** [review]
👍🏻LGTM
